src/flask_pyoidc/flask_pyoidc.py

```diff
@@ -222,6 +233,18 @@ def wrapper(*args, **kwargs):
                 elif session.is_authenticated():
                     logger.debug('user is already authenticated')
                     return view_func(*args, **kwargs)
+                # If the user is not authenticated, check if the user has
+                # passed access token in the request header and make
+                # introspection request to the identity provider to verify the
+                # token.
+                elif accept_token and self._check_authorization_header(flask.request):
+                    if self.introspect_token(request=flask.request,
+                                             client=client, scopes=scopes_required):
+                        logger.info('user has valid access token')
+                        return view_func(*args, **kwargs)
```

Comment:
This should pass some of the introspected token data to the `view_func`, like the `sub` claim if it's in the token and any client_id, to allow the `view_func` to determine which client/user is calling the endpoint.

This would aslo be yet another reason to have it as separate decorator, since it would pass som arguments to the decorated view function.

Reply:
If the access_token is shared by the service to other service, would that be useful for the `view_func` to know about the token?

Reply:
What do you mean with 
> If the access_token is shared by the service to other service

In the case of token introspection here, I thought it would be for the app to be able to receive tokens for authorization.
If it doesn't suffice to know that a token was received, the `view_func` also needs to know *which* user the token was issued for (to be able to read any data related to that user).

Reply:
In service to service architecture, Bob's service requests access token from the Identity Provider using Client Credentials grant. This access token can be used for two things:

1. Bob's service can access its own API using that token.
2. Bob's service can share that token to Alice's service so that her service can also access Bob's API. Bob's service can verify that token using token introspection. Only Bob can verify that token because it requires client_id and client_secret to make token introspection request.

So if the owner of that token is Bob, the view function of Bob's service will always see the userinfo of Bob which is not useful.

Reply:
With the service-to-service case: it's typically not the same client that introspects the token as the one that it was issued to.
What happens is usually:
1. Bob's service wants to access Alice's service, and obtains an access token from a common provider using Bob's client credentials.
2. Bob's service passes the token with the request to Alice's service.
3. Alice's service sends an introspection request with the token to the provider using Alice's client credentials.
4. Optional: the provide may restrict which client can introspect which tokens
5. Alice's service gets back the introspection response which identifies Bob's service as the "token owner" and may reject or accept Bob's service request.

The token might also be an access token issued for a user that has logged-in to Bob's service, i.e. issued after Authorization Code Flow (and not client credentials grant).
Hence it's necessary to expose the introspection information - it might be issued to a user and probably to another client than the service receiving it.

Reply:
Resolved. Now `introspect_token` returns entire token info containing information about the user as well.